controller/api/destination/external-workload/endpoints_controller.go

```diff
@@ -0,0 +1,513 @@
+					}
+
+					for k, v := range oldServices {
+						if _, ok := services[k]; !ok {
+							disjoint[k] = v
+						}
+					}
+
+					services = disjoint
+				}
+			}
+
+			for svc := range services {
+				ec.enqueueUpdate(svc)
+			}
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Store configuration for leader elector client. The leader elector will
+	// accept three callbacks. When a lease is claimed, the elector will mark
+	// the manager as a 'leader'. When a lease is released, the elector will set
+	// the isLeader value back to false.
+	ec.lec = leaderelection.LeaderElectionConfig{
+		// When runtime context is cancelled, lock will be released. Implies any
+		// code guarded by the lease _must_ finish before cancelling.
+		ReleaseOnCancel: true,
+		Lock: &resourcelock.LeaseLock{
+			LeaseMeta: metav1.ObjectMeta{
+				Name:      leaseName,
+				Namespace: controllerNs,
+			},
+			Client: k8sAPI.Client.CoordinationV1(),
+			LockConfig: resourcelock.ResourceLockConfig{
+				Identity: hostname,
+			},
+		},
+		LeaseDuration: leaseDuration,
+		RenewDeadline: leaseRenewDeadline,
+		RetryPeriod:   leaseRetryPeriod,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(context.Context) {
+				ec.Lock()
+				defer ec.Unlock()
+				ec.isLeader = true
+			},
+			OnStoppedLeading: func() {
+				ec.Lock()
+				defer ec.Unlock()
+				ec.isLeader = false
+				ec.log.Infof("%s released lease", hostname)
+			},
+			OnNewLeader: func(identity string) {
+				if identity == hostname {
+					ec.log.Infof("%s acquired lease", hostname)
+				}
+			},
+		},
+	}
+
+	return ec, nil
+}
+
+// Start will run the endpoint manager's processing loop and leader elector.
+//
+// The function will spawn two background tasks; one to run the leader elector
+// client that and one that will process updates applied by the informer
+// callbacks.
+func (ec *EndpointsController) Start() {
+	// Create a parent context that will be used by leader elector to gracefully
+	// shutdown.
+	//
+	// When cancelled (either through cancel function or by having its done
+	// channel closed), the leader elector will release the lease and stop its
+	// execution.
+	ctx, cancel := context.WithCancel(context.Background())
+
+	go func() {
+		for {
+			// Block until a lease is acquired or a lease has been released
+			leaderelection.RunOrDie(ctx, ec.lec)
+			// If the context has been cancelled, exit the function, otherwise
+			// continue spinning.
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+		}
+	}()
+
+	// Start a background task to process updates. When a shutdown signal is
+	// received over the manager's stop channel, it is propagated to the elector
+	// through the context object to ensure the elector task does not leak.
+	go func() {
+		for {
+			select {
+			case update := <-ec.updates.queue:
+				elapsed := time.Since(update.enqueTime).Seconds()
+				ec.updates.queueLatency.Observe(elapsed)
+				ec.updates.queueLength.Dec()
```

Comment:
Can we register a gauge function that returns `len(ec.updates.queue)` instead of managing the gauge value manually?

Reply:
I think that is right. Each time you touch the queue, just measure the len